File: kotlin/space-age/src/main/kotlin/SpaceAge.kt
import Planet.*

class SpaceAge(val seconds: Long) {

    private val planetRatio = hashMapOf(
        Mercury to 0.2408467,
        Venus to 0.61519726,
        Earth to 1.0,
        Mars to 1.8808158
    )
            
    fun onEarth(): Double = on(Earth)

    fun onMercury(): Double = on(Mercury)

    fun onVenus(): Double = on(Venus)

    fun onMars(): Double = on(Mars)

    private fun on(planet: Planet): Double {
        val ratio = planetRatio.getValue(planet)
        val yearsOnEarth = seconds / (365.25 * 24 * 60 * 60)
        return roundTwoDecimals(yearsOnEarth / ratio)
    }

    fun onNeptune(): Double {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    fun onUranus(): Double {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    fun onSaturn(): Double {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    fun onJupiter(): Double {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    private fun roundTwoDecimals(value: Double) = Math.round(value * 100) / 100.0
}

File: haskell/space-age/src/SpaceAge.hs
module SpaceAge
  ( Planet(..)
  , ageOn
  ) where

data Planet
  = Mercury
  | Venus
  | Earth
  | Mars
  | Jupiter
  | Saturn
  | Uranus
  | Neptune

ageOn :: Planet -> (Float -> Float)
ageOn planet =
  let earthYearInSeconds = 31557600
   in (/ (earthYearInSeconds * orbitalRelativeToEarth planet))

orbitalRelativeToEarth :: Planet -> Float
orbitalRelativeToEarth planet =
  case planet of
    Earth -> 1
    Mercury -> 0.2408467
    Venus -> 0.61519726
    Mars -> 1.8808158
    Jupiter -> 11.862615
    Saturn -> 29.447498
    Uranus -> 84.016846
    Neptune -> 164.79132
